cmd/lhsm-plugin-az/mover.go

```diff
@@ -86,30 +88,62 @@ func (m *Mover) fileIDtoContainerPath(fileID string) (string, string, error) {
 	return container, path, nil
 }
 
+func (m *Mover) getSASToken() string {
+	m.config.configLock.Lock()
+	defer m.config.configLock.Unlock()
+	return m.config.AzStorageSAS
+}
+
 func (m *Mover) refreshCredential() {
 	const sigAzure = "sig="
-	var err error
-	if m.nextCredRefreshTime.After(time.Now()) {
-		return
-	}
-
-	m.cred = azblob.NewAnonymousCredential()
-	sas, err := util.GetKVSecret(m.config.AzStorageKVName, m.config.AzStorageKVSecretName)
-
-	if err != nil {
-		util.Log(pipeline.LogError, fmt.Sprintf("Failed to update SAS. Falling back to previous SAS.\n%s", err))
-		return
-	}
-
-	if !strings.Contains(sas, sigAzure) {
-		util.Log(pipeline.LogError, fmt.Sprintf("Failed to update SAS, invalid SAS returned. Falling back to previous SAS."))
-		return
+	defaultRefreshInterval, _ := time.ParseDuration(m.config.CredRefreshInterval)
+	nextRefreshInterval := defaultRefreshInterval
+	retryDelay := 4*time.Second // 4 seconds
+	maxTries := 20
+	try := 1 // Number of retries to get the SAS. Starts at 1.
+
+	for {
+		select {
+		case <- time.After(nextRefreshInterval):
+			sas, err := util.GetKVSecret(m.config.AzStorageKVName, m.config.AzStorageKVSecretName)
+
+			if err == nil && !strings.Contains(sas, sigAzure) {
+				err = fmt.Errorf("invalid SAS returned")
+			}
+
+			if err != nil {
+
+				util.Log(pipeline.LogError,	fmt.Sprintf(
+							"Failed to update SAS. Falling back to previous SAS.\nReason: %s, try: %d",
+							err, try))
+				
+				/* 
+				 * Failed to update SAS. We'll retry with exponential delay.
+				 */
+				if (try <= maxTries) {
+					nextRefreshInterval = time.Duration(math.Pow(2, float64(try))-1) * retryDelay
+					try++
+				} else {
+					util.Log(pipeline.LogError, fmt.Sprintf("Could not update SAS in %d attempts. Giving up.", try))
+					nextRefreshInterval = defaultRefreshInterval
```

Comment:
If I understood correctly, we'll give up and wait for the next 24H. Hmmm, this is indeed a difficult situation to be in.